superset/databases/dao.py

```diff
@@ -0,0 +1,75 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import logging
+from contextlib import closing
+
+from flask import g
+from sqlalchemy import select
+
+from superset import app
+from superset.dao.base import BaseDAO
+from superset.extensions import db
+from superset.models.core import Database
+from superset.security.analytics_db_safety import check_sqlalchemy_uri
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseDAO(BaseDAO):
+    model_cls = Database
+
+    @staticmethod
+    def test_connection(  # pylint: disable=too-many-arguments, too-many-return-statements
+        db_name: str,
+        uri: str,
+        server_cert: str,
+        extra: str,
+        impersonate_user: bool,
+        encrypted_extra: str,
+    ) -> None:
+        if app.config["PREVENT_UNSAFE_DB_CONNECTIONS"]:
+            check_sqlalchemy_uri(uri)
+        # if the database already exists in the database, only its safe
+        # (password-masked) URI would be shown in the UI and would be passed in the
+        # form data so if the database already exists and the form was submitted
+        # with the safe URI, we assume we should retrieve the decrypted URI to test
+        # the connection.
+        if db_name:
+            existing_database = (
+                db.session.query(Database)
+                .filter_by(database_name=db_name)
+                .one_or_none()
+            )
+            if existing_database and uri == existing_database.safe_sqlalchemy_uri():
+                uri = existing_database.sqlalchemy_uri_decrypted
+
+        # this is the database instance that will be tested
+        database = Database(
```

Comment:
Here through line 70 could also potentially be a DAO method - something like `build_db_for_connection_test`